File: RibbonFactory/Enums/Enumeration.vb
Imports System.Reflection

Namespace Enums

    Public MustInherit Class Enumeration
        Implements IEquatable(Of Enumeration)
        Implements IComparable(Of Enumeration)
        Implements ICloneable

        Protected ReadOnly value As Integer
        Protected ReadOnly name As String

        Protected Sub New(value As Integer, name As String)
            Me.value = value
            Me.name = name
        End Sub

        Public Shared Function Values(Of T As Enumeration)() As T()
            Return GetType(T).
                GetProperties(BindingFlags.Public Or BindingFlags.Static Or BindingFlags.DeclaredOnly).
                Select(Function(p) p.GetValue(Nothing)).
                OfType(Of T)().
                ToArray()
        End Function

        Public Shared Function Parse(Of T As Enumeration)(value As Integer) As T
            Return Values(Of T)().Single(Function(v) v.value = value)
        End Function

        Public Shared Function Parse(Of T As Enumeration)(name As String) As T
            Return Values(Of T)().Single(Function(v) String.Equals(v.name, name, StringComparison.OrdinalIgnoreCase))
        End Function

        Public MustOverride Function Clone() As Object Implements ICloneable.Clone

#Region "Overrides And Comparison"

        Public Overrides Function ToString() As String
            Return name
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return value
        End Function

        Public Overrides Function Equals(obj As Object) As Boolean
            Return Equals(TryCast(obj, Enumeration))
        End Function

        Public Overloads Function Equals(other As Enumeration) As Boolean Implements IEquatable(Of Enumeration).Equals
            Return other IsNot Nothing AndAlso other.value = value
        End Function

        Public Function CompareTo(other As Enumeration) As Integer Implements IComparable(Of Enumeration).CompareTo
            Return value.CompareTo(other.value)
        End Function

#End Region

#Region "Operators"

        Public Shared Operator =(left As Enumeration, right As Enumeration) As Boolean
            If left Is Nothing AndAlso right Is Nothing Then
                Return True
            ElseIf left Is Nothing OrElse right Is Nothing Then
                Return False
            Else
                Return left.Equals(right)
            End If
        End Operator

        Public Shared Operator <>(left As Enumeration, right As Enumeration) As Boolean
            Return Not left = right
        End Operator

        Public Shared Operator >(left As Enumeration, right As Enumeration) As Boolean
            Return left.CompareTo(right) > 0
        End Operator

        Public Shared Operator <(left As Enumeration, right As Enumeration) As Boolean
            Return left.CompareTo(right) < 0
        End Operator

        Public Shared Operator >=(left As Enumeration, right As Enumeration) As Boolean
            Return left.CompareTo(right) >= 0
        End Operator

        Public Shared Operator <=(left As Enumeration, right As Enumeration) As Boolean
            Return left.CompareTo(right) <= 0
        End Operator

#End Region

    End Class

End Namespace

File: RegistrationUtility/ModelObjects/Enumeration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace RegistrationUtility.ModelObjects
{
    internal abstract class Enumeration : IEquatable<Enumeration>, IComparable<Enumeration>
    {
        protected readonly int value;
        protected readonly string name;

        protected Enumeration(int value, string name)
        {
            this.value = value;
            this.name = name;
        }

        public static T[] Values<T>() where T : Enumeration
        {
            return typeof(T)
                .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
                .Select(p => p.GetValue(null))
                .OfType<T>()
                .ToArray();
        }

        public static T Parse<T>(int value) where T : Enumeration
        {
            return Values<T>().Single(v => v.value == value);
        }

        public static T Parse<T>(string name) where T : Enumeration
        {
            return Values<T>().Single(v => string.Equals(v.name, name, StringComparison.OrdinalIgnoreCase));
        }

        #region Overrides And Comparison

        public override string ToString()
        {
            return $"({value}) {name}";
        }

        public override int GetHashCode()
        {
            return value;
        }

        public override bool Equals(object obj)
        {
            return Equals(obj as Enumeration);
        }

        public bool Equals(Enumeration other)
        {
            return other != null && value == other.value;
        }

        public int CompareTo(Enumeration other)
        {
            return value.CompareTo(other?.value);
        }

        #endregion

        #region Operators

        public static bool operator ==(Enumeration left, Enumeration right)
        {
            return EqualityComparer<Enumeration>.Default.Equals(left, right);
        }

        public static bool operator !=(Enumeration left, Enumeration right)
        {
            return !(left == right);
        }

        public static bool operator <(Enumeration left, Enumeration right)
        {
            return left.CompareTo(right) < 0;
        }

        public static bool operator <=(Enumeration left, Enumeration right)
        {
            return left.CompareTo(right) <= 0;
        }

        public static bool operator >(Enumeration left, Enumeration right)
        {
            return left.CompareTo(right) > 0;
        }

        public static bool operator >=(Enumeration left, Enumeration right)
        {
            return left.CompareTo(right) >= 0;
        }

        #endregion

    }
}
